woocommerce/woocommerce-gutenberg-products-block
# Try move shipping related data to a `@wordpress/data` store

This PR depends on #5981 - the changes from that are included in this PR.

This PR will create a new hook called `useShippingData` it is very similar to the extant `useShippingDataContext` but instead of getting data from the context provider, the data is derived from the `wc/store/v1/cart` data store. You can get this hook from `@woocommerce/base-context/hooks`.

At the moment, the `shippingErrorStatus` `dispatchErrorStatus`, `shippingErrorTypes` and the event observers are still kept in context as it doesn't make sense to move them outside of it.

- Updated tests to use `useShippingData` rather than `useShippingDatacontext`.
- Create a hook `useShippingData` that gets shipping data out of the `wc/store/v1/cart` data store.
- Update `assets/js/base/context/hooks/payment-methods/use-payment-method-interface.ts` because it was returning incorrect methods. There is no such thing as `setSelectedRates`, it is `selectShippingRate`.
- Update types in `assets/js/types/type-defs/payment-method-interface.ts`.
- Create `SelectShippingRateType` which describes the shape of the value returned by the `useSelectShippingRate` hook.
- Convert `assets/js/base/utils/derive-selected-shipping-rates.ts` to TypeScript.
- Add the following selectors to the `wc/store/v1/cart` store:
   - `getShippingRates`
   - `getNeedsShipping`
   - `getHasCalculatedShipping`


Fixes #5898 




### Testing

#### Automated Tests
### Manual Testing

How to test the changes in this Pull Request:

1. Set up multiple shipping methods, ensure you can change between them. Ensure the price updates when selecting different ones.
2. Check out using a few different payment methods, make sure to try express payment methods and ensure the correct address is submitted.
3. Ensure CI tests pass.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).



### Changelog

> Add suggested changelog entry here.
